### ISSUE TITLE
Force id to be numerical

### ISSUE THREAD

**Issue by waghanza** [the-benchmarker/web-frameworks#1148]
Hi,

Some implementations here **SHOULD** be considered as `unsecure`

`/users/foo1` could lead to a successful HTTP query

Regards,

**Comment by waghanza** [comment]
@ohler55 Could a regexp be used on `agoo-c` ? (or any way to force a constraint)

**Comment by ohler55** [comment]
Are you referring to the handler path setup? If so you can set up path for explicit files or for directories. The current example in the benchmark handles any path but it could be made explicit if that is the desired behaviour.

**Comment by waghanza** [comment]
@mfirry Thanks for reply on https://github.com/the-benchmarker/web-frameworks/pull/1147

I find a way for `aspnetcore` :stuck_out_tongue: Is there a way to force an URL param to be a numeric - > having a route that match `/user/1` but not `/user/a`

**Comment by mfirry** [comment]
Working on it. Will submit a PR for both akka-http and http4s. Is it ok?

**Comment by waghanza** [comment]
Yes. I'll work on your `PR` then un order to homogenize behavior for all implementations here

**Comment by ohler55** [comment]
It will take some time to get those features in. TLS is on my roadmap but I hadn't considered akka-http.

As for the filtering of URL paths, I can see a desire on the part of the application to filter with finer granularity but I don't see how it could be perceived as a security issue. After all, the application just needs to look at the path tail (number or string) and decide whether to process or not. Since the logic could be as complex as "only accept prime numbers" it doesn't seem like something that needs to be pushed into the server. It becomes an application decision at some point.

**Comment by waghanza** [comment]
@ohler55 
> It does handle single path component matching and "rest" of path matching though

Can `agoo-c` handle a dynamic path (but not `*`) ?

**Comment by waghanza** [comment]
@ohler55 This is not for security only, but more to homogenize behaviour between each frameworks

**Comment by ohler55** [comment]
Ahh, okay, the start of the issue suggested handling the `/user/foo1` was unsecure. I'll ignore that comment then.

Agoo supports two pattern matching strings for server filtering as apposed to application filtering. The `*` character matches exactly one path element where path elements are the parts of the path delimited by the `/` character. So `/user/*`will match `/user/foo` but not `/user/foo/1`. A sequence of `**` will match the rest of the path and only is accepted at the end of the pattern so `/user/**` will match `user/1` and `user/foo/1`.

I suspect it is rare to find support for regex support in paths. Rake for example does not support it but expects the application to handle all routing. How do you expect the frameworks to deal with this new requirement? Will it be okay to just write a bit of code in the app or will all servers that don't support the feature be marked as failing?

**Comment by waghanza** [comment]
`rails` support adding a constraint to route 
~~~ruby
match 'maps/:id', to: 'maps#show' ,constraints: { id: /\d+/}, via: :get
~~~

**Comment by waghanza** [comment]
> So, whats the plan then? A requirement to be on the list? Rails is of course a layer under a server. 

I do not see think like that. For me it's more a great step forward to make frameworks uniform here, since :
+ it's already implemented in somes (natively)
+ it's seems to be implemented like that in **production** app

> Sounds like it that is acceptable then implementing the functionality in the app would be acceptable, yes?

Yes

**Comment by ohler55** [comment]
As long as there is a short term option to filter in the app I can get that done fairly quickly.
